---
title: "processgeuvadis.Rmd"
output: html_document
author: Jeff Leek
---

`r library(knitr); opts_chunk$set(cache=TRUE)`

### Load packages

You will need the RSkittleBrewer package for this vignette to run. Installation instructions are available here:

* https://github.com/alyssafrazee/RskittleBrewer

You will also need the "for-bioc" branch of the Ballgown package available for install from here: 

https://github.com/alyssafrazee/ballgown/tree/alpha

And the processed GEUVADIS fpkm data from here: 

https://github.com/alyssafrazee/ballgown_code

You will also need R version 3.1.0 or greater and Bioconductor 3.0 or greater. The zebrafishRNASeq package might need to be installed from source. These analyses are based on the devel version of sva (version 3.11.2 or greater).

```{r load,message=FALSE}
library(zebrafishRNASeq)
library(RSkittleBrewer)
library(genefilter)
library(Biobase)
library(ballgown)
library(RUVSeq)
library(edgeR)
library(sva)
library(ffpe)
library(RColorBrewer)
library(corrplot)
library(limma)
trop = RSkittleBrewer('tropical')
```


## Load the data

You will need to download the GEUVADIS ballgown object from this site: https://github.com/alyssafrazee/ballgown_code


```{r loaddata,dependson="load"}
load("fpkm.rda")
pd = ballgown::pData(fpkm)
pd$dirname = as.character(pd$dirname)
ss = function(x, pattern, slot=1,...) sapply(strsplit(x,pattern,...), "[", slot)
pd$IndividualID = ss(pd$dirname, "_", 1)
tfpkm = expr(fpkm)$trans
```

## Subset to non-duplicates

You will need the GEUVADIS quality control information and population information available from these sites: https://www.dropbox.com/s/rg63qtuws2liz9r/GD667.QCstats.masterfile.txt, https://www.dropbox.com/s/woacfjxql7gxhnt/pop_data_withuniqueid.txt.

```{r pheno,dependson="loaddata"}
pheno = read.delim("GD667.QCstats.masterfile.txt", as.is=TRUE)
m = read.delim("pop_data_withuniqueid.txt",as.is=TRUE)
pd$SampleID = m$sample_id[match(pd$dirname, m$folder_id)]
pd$UseThisDup = pheno$UseThisDuplicate[match(pd$SampleID, rownames(pheno))]
pd$batch = pheno$RNAExtractionBatch[match(pd$SampleID, rownames(pheno))]
pd$lab = pheno$SeqLabNumber[match(pd$SampleID, rownames(pheno))]
pd$libprepdate = pheno$LibraryPrepDate[match(pd$SampleID, rownames(pheno))]
## drop duplicates for this

pd = pd[pd$UseThisDup == 1,]

## subset the transcript expression data to match pheno data

colnames(tfpkm) = ss(colnames(tfpkm),"\\.",2)
tfpkm = tfpkm[,pd$dirname]

## Remove low expression transcripts
rowmfpkm = rowMeans(log(tfpkm+1)) 
keepIndex=which(rowmfpkm > 1)
tfpkm = tfpkm[keepIndex,]
tfpkm = as.matrix(tfpkm)
```

## Run the different analyses and compare to lab

```{r estbatch1,dependson="pheno"}
mod1 = model.matrix(~ pd$population)
mod0 = cbind(mod1[,1])

## Estimate batch with svaseq (unsupervised)
batch_unsup_sva = svaseq(tfpkm,mod1,mod0,n.sv=1)$sv
```


```{r estbatch2, dependson="estbatch1"}
## Estimate batch with pca
ldat0 = log(tfpkm + 1)
batch_pca = svd(ldat0 - rowMeans(ldat0))$v[,1]


```

```{r estbatch3,dependson="estbatch3"}
## Estimate batch with ruv (residuals)
## this procedure follows the RUVSeq vignette
## http://www.bioconductor.org/packages/devel/bioc/vignettes/RUVSeq/inst/doc/RUVSeq.pdf

x <- as.factor(pd$population)
design <- model.matrix(~x)
y <- DGEList(counts=tfpkm, group=x)
y <- calcNormFactors(y, method="upperquartile")
y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)
fit <- glmFit(y, design)
res <- residuals(fit, type="deviance")
seqUQ <- betweenLaneNormalization(tfpkm, which="upper")
controls = rep(TRUE,dim(tfpkm)[1])
batch_ruv_res = RUVr(seqUQ,controls,k=1,res)$W
```

```{r estbatch4, dependson="estbatch3"}

## Estimate batch with ruv empirical controls
## this procedure follows the RUVSeq vignette
## http://www.bioconductor.org/packages/devel/bioc/vignettes/RUVSeq/inst/doc/RUVSeq.pdf

y <- DGEList(counts=tfpkm, group=x)
y <- calcNormFactors(y, method="upperquartile")
y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)

fit <- glmFit(y, design)
lrt <- glmLRT(fit, coef=2)

controls = rank(lrt$table$LR) <= 400
batch_ruv_emp <- RUVg(tfpkm, controls, k=1)$W
```

## Correlation between estimates

```{r corr,dependson="estbatch4"}
batchEstimates = cbind(batch_unsup_sva,batch_pca,batch_ruv_res,batch_ruv_emp)
colnames(batchEstimates) = c("usva","pca","ruvres","ruvemp")

corr = abs(cor(batchEstimates))
cols = colorRampPalette(c(trop[2],"white",trop[1]))
corrplot(corr,method="ellipse",type="lower",col=cols(100),tl.pos="d")
```



## Boxplot the results by lab


```{r plotResults,dependson="estbatch4"}
boxplot(batch_unsup_sva ~ pd$lab,ylab="Unsupervised SVA",col=trop[1])
boxplot(batch_pca ~ pd$lab,ylab="Unsupervised PCA",col=trop[2])
boxplot(batch_ruv_res ~ pd$lab,ylab="Unsupervised RUV residuals",col=trop[3])
boxplot(batch_ruv_emp ~ pd$lab,ylab="Unsupervised RUV control probes",col=trop[4])
```

## ANOVA between estimates and lab


```{r summarizeResults,dependson="estbatch4"}
anova(lm(batch_unsup_sva ~ pd$lab))
anova(lm(batch_pca ~ pd$lab))
anova(lm(batch_ruv_res~ pd$lab))
anova(lm(batch_ruv_emp~ pd$lab))
```


## Differential expression results

```{r comparede, dependson="estbatch4",fig.align="center",fig.height=7,fig.width=7}
dge <- DGEList(counts=tfpkm)
dge <- calcNormFactors(dge)
catplots = tstats = vector("list",6)
adj = c("+ pd$lab","+ batch_pca", "+ batch_unsup_sva",
        "+ batch_ruv_res", "+ batch_ruv_emp", "")

for(i in 1:6){
  design = model.matrix(as.formula(paste0("~ pd$population",adj[i])))
  v <- voom(dge,design,plot=FALSE)
  fit <- lmFit(v,design)
  fit <- eBayes(fit)
  tstats[[i]] = abs(fit$t[,2])
  names(tstats[[i]]) = 1:length(tstats[[i]])
  catplots[[i]] = CATplot(-rank(tstats[[i]]),-rank(tstats[[1]]),maxrank=3000,make.plot=F)
  cat(i)
}

plot(catplots[[2]],ylim=c(0,1),col=trop[1],lwd=3,type="l",ylab="Concordance Between True rank and rank with different methods",xlab="Rank")
lines(catplots[[3]],col=trop[2],lwd=3)
lines(catplots[[4]],col=trop[3],lwd=3,lty=2)
lines(catplots[[5]],col=trop[3],lwd=3,lty=1)
lines(catplots[[6]],col=trop[4],lwd=3,lty=2)


legend(200,0.5,legend=c("PCA", "Unsup. svaseq","RUV Res.", "RUV Emp.", "No adjustment"),col=trop[c(1,2,3,3,4)],lty=c(1,1,2,1,2),lwd=3)
```




## Session Information


```{r}
sessionInfo()
```



